### ISSUE TITLE
Possible refactors

### ISSUE THREAD

**Issue by jat113** [TJenks13/StreamVideo-1#22]
I made this comment about possible refactors in a previous issue but I'm making another issue with the label enhancement since the other issue was only intended for deserializing the video.csv file.

So with that out of the way here's my thoughts about refactors:

1. Can we improve the construction of streams in the function createStream in StreamingActivity?
2.  Can we improve the construction of videos in the video class constructor using a DataRecord? I'm thinking an extract method on the if statements to return the type.
3. I think that the entire function report in Account has a lot of parameters. 
4. I think that in the video class we can create a structure like RunningTime that will contain the hours and minutes of the video. Or just make it a single field with the time in minutes and we can create a function to retrieve the minutes and hours from it.

**Comment by jat113** [comment]
I was thinking about this more. I reverted back to how files were being loaded in StreamingActivity and deleted the DataRecord files so there's no longer a refactor for 2. However we can still do 3, and 4. And I think the current factory method we have now for constructing streams is probably as good as it is going to get.

**Comment by henry-nnnn** [comment]
3. Account.cpp? Do you mean the total variables? We could make a class to hold the totals.

4. i thought about that too because we can just use simple math to retrieve the hours by dividing minute, but i was weary on that as I think it might be creating new functions?

**Comment by henry-nnnn** [comment]
Actually, I'm technically working on 4 right now, but I'm having a slight issue with trying to bring it down to 4 arguments with the "RunningTime" object instead while not breaking the program that initially took 5 arguments in the constructor. If that's where I think we're both going with it?

**Comment by TJenks13** [comment]
We can create new functions to make easier writing of test cases, or classes, but creating that sort of function sounds like it may be getting into new functionality territory.

**Comment by TJenks13** [comment]
For 3, like Henry said, we could just create a new class and shift all of the totals variables out of it?

**Comment by jat113** [comment]
I can see why you guys say it might be getting into new functionality. I'm thinking however that as long as the refactor does the exact same thing as the original code did, then I don't think that would be creating  new functionality. I certainly think we can group minutes and hours into some structure and it might be best to keep it in the same format. But one thing that annoys me is how the report function calculates the total time. Maybe we ought to create a method in the stream which returns the total time with the data already formatted appropriately.

**Comment by jat113** [comment]
the report function in account that is.

**Comment by TJenks13** [comment]
I agree, that could be of use. Maybe we could remove all of the data variables from the account.cpp function report and place them in a structure, extracting the primitives from the function and then having a function which does the minutes and hours calculations itself.

**Comment by jat113** [comment]
> Actually, I'm technically working on 4 right now, but I'm having a slight issue with trying to bring it down to 4 arguments with the "RunningTime" object instead while not breaking the program that initially took 5 arguments in the constructor. If that's where I think we're both going with it?

If you are having difficulties with creating the RunningTime object and getting the program to build, I can do that for you. I already have it done in a local repository, but I wasn't going to push it since you created the issue for video primitives in the first place. 

Also looking at account report, I think the way it calculates the total stream events and the non-original stream events is confusing. Not only that, I don't even know what a "stream event" or "non-original stream event" is because the function is doing so many weird things. If its a movie it checks how many times it "occurred" which I think this means watched, and then multiplies it by the number of hours; If it is a show, or an original it checks how many times it "occurred." 

Another problem I have with this code is that I haven't the slightest idea what an "original" video even means but I do know its been a pain in my ass ever since I replaced the conditionals with polymorphism. 

Also I'm going to rename the member occurrences and getOccurrences in the Stream class to watched and timesWatched respectively.

**Comment by jat113** [comment]
Also it just occurred to me that there's no way that account should ever need to call get occurrences to get the totalhours or minutes. That's something that should be in the stream and I'm going to move that into the stream.

**Comment by henry-nnnn** [comment]
Could you set a snippet of your approach in a branch for that RunTime? And I was originally trying to make the RunTime class do the calculations to convert the minutes (RunTime should just hold minutes was what I was going for) and create a function that converts the minutes to hours so that we can move that function from the Account into the RunTime to avoid the RunTime from smelling like just a Data Type Class

**Comment by henry-nnnn** [comment]
Also I brought the issue of the setLength function on how it was unused and should be deleted

**Comment by jat113** [comment]
I also think we can remove setlength so that's a good issue.

As far as how I managed to group the primitives hours and minutes into a class, I just created a source and header file containing the class, added the necessary includes in the files, and replaced the primitives with the class. Although, if you add a file, you got to change the makefile. If you don't add any files just put the class somewhere and include the file when you need it.

**Comment by henry-nnnn** [comment]
I created branch issue16 for the class issue (branch15 was already used/named incorrect issue) and uploaded what I got. I need some assistance taking the values from the Video constructor and passing it into the RunTime class.

**Comment by jat113** [comment]
Ok one moment I'll make a branch with the class in it.

**Comment by jat113** [comment]
I will create the class and that way you can refactor the primitives

**Comment by henry-nnnn** [comment]
Okay! Thanks So Much. I'll try my best

**Comment by jat113** [comment]
Sorry  it took me so long. I was having trouble adding it.

**Comment by jat113** [comment]
You will have to change a lot of test cases to get it added however. Since you will need to change all the places where the video class is constructed.

**Comment by jat113** [comment]
If there's any other refactors you would like to do on the code you can do that and I can finish it. Up to you though.

**Comment by henry-nnnn** [comment]
Yeah. okay. my main problems was literally to try to do it in a way where I didn't have to change all the Test Code involving the Video constructor to take the object RunTime instead. I guess there wasn't way around it haha

**Comment by henry-nnnn** [comment]
we would essentially change the push.back function too in the main SteamActivity.cpp?

**Comment by jat113** [comment]
Yes. And yes there isn't any other way to get around the test cases

**Comment by henry-nnnn** [comment]
I finished refactoring. and yeah haha there's a lot of places i have to redo the video constructors

**Comment by henry-nnnn** [comment]
i pushed the previous branch with the refactored but would have still change all the video constructors
